src/main/java/com/ibm/eventstreams/connect/mqsource/builders/BaseRecordBuilder.java

```diff
@@ -160,38 +209,219 @@ public abstract SchemaAndValue getValue(JMSContext context, String topic, boolea
-                    key.value(),
-                    value.schema(),
-                    value.value()
-            );
+                    headers);
+        } catch (final Exception dlqException) {
+            // If DLQ processing itself fails, log and skip
+            log.error("Failed to create DLQ record: {}", dlqException.getMessage(), dlqException);
+            return null;
         }
     }
+
+    /**
+     *
+     * Extracts payload from a JMS message with improved error handling
+     *
+     * @param message The actual message coming from mq
+     *
+     * @return Optional<byte[]>
+     */
+    private Optional<byte[]> extractPayload(final Message message) {
+        try {
+            if (message instanceof BytesMessage) {
+                return Optional.ofNullable(message.getBody(byte[].class));
+            } else if (message instanceof TextMessage) {
+                final String text = message.getBody(String.class);
+                return Optional.ofNullable(text != null ? text.getBytes(UTF_8) : null);
+            } else {
+                log.error("Unsupported JMS message type {} for DLQ", message.getClass().getName());
+                return Optional.empty();
+            }
+        } catch (final JMSException e) {
+            log.error("Failed to extract payload: {}", e.getMessage());
+            return Optional.empty();
+        }
+    }
+
+    /**
+     *
+     * Creates enhanced headers with error information for DLQ records
+     *
+     * @param originalTopic The original topic name
+     * @param exception     The execption that needs to be included in the header
+     *
+     * @return Headers
+     */
+    private Headers createErrorHeaders(final String originalTopic, final Exception exception) {
+        final Headers headers = new ConnectHeaders();
+
+        // Basic error information
+        headers.addString("original_topic", originalTopic);
+        headers.addString("error_message", exception.getMessage());
+        headers.addString("error_class", exception.getClass().getName());
+        headers.addString("error_timestamp", String.valueOf(System.currentTimeMillis()));
+
+        // Add cause if available
+        if (exception.getCause() != null) {
+            headers.addString("error_cause", exception.getCause().getMessage());
+            headers.addString("error_cause_class", exception.getCause().getClass().getName());
+        }
+
+        // Add first few lines of stack trace (full stack trace might be too large)
+        try {
+            final StringWriter sw = new StringWriter();
+            final PrintWriter pw = new PrintWriter(sw);
+            exception.printStackTrace(pw);
+            final String stackTrace = sw.toString();
+
+            // First 500 characters or less to avoid overly large headers
+            final String truncatedStackTrace = stackTrace.length() <= 500 ? stackTrace
+                    : stackTrace.substring(0, 500) + "... [truncated]";
```

Comment:
I suggest moving this into a separate method for getting the stack trace from an exception